File: web/test/integration/test/FileUploadSpec.js
'use strict'

const asUser = require('../support/asUser')
const shouldBehaveLikeATree = require('../support/behave/likeATree')

describe('FileUpload', function() {
  asUser.usingTemporaryUser(function() {
    before(function() {
      this.browser.loadShortcuts('documentSets')
      this.browser.loadShortcuts('importFiles')

      this.b = this.browser
      this.documentSets = this.b.shortcuts.documentSets
      this.importFiles = this.b.shortcuts.importFiles
    })

    describe('after creating a file with metadata', function() {
      before(async function() {
        // Add a file, with metadata 'foo': 'bar'
        await this.importFiles.open()
        await this.importFiles.addFiles(['FileUpload/Cat1.docx'])
        await this.b.click({ button: 'Done adding files' })

        await this.b.sleep(500) // wait for dialog to appear
        await this.b.sendKeys('metadata-test', { tag: 'input', name: 'name', wait: true })
        await this.b.click({ link: 'Add new field' })
        await this.b.sendKeys('foo', { tag: 'input', id: 'document-metadata-add-field' })
        await this.b.click('#document-metadata-add-field + button')
        await this.b.sendKeys('bar', { tag: 'input', name: 'foo' })
        await this.b.click({ button: 'Import documents' })
        await this.importFiles.waitUntilRedirectToDocumentSet('metadata-test')
      })

      it('should add metadata to the imported file', async function() {
        await this.b.click({ tag: 'h3', contains: 'Cat1.docx' })

        await this.b.sleep(1000) // wait for document to appear
        await this.b.click({ link: 'Fields', wait: true })

        // wait for metadata to appear
        const value = await this.b.getAttribute({ tag: 'input', name: 'foo', wait: true }, 'value')
        expect(value).to.eq('bar')
      })

      it('should add more metadata in a second import', async function() {
        // Add another file, with metadata 'moo':'mar'
        await this.b.click([ { tag: 'nav' }, { link: 'metadata-test' } ])
        await this.b.click({ link: 'Add Documents' })
        await this.importFiles.waitUntilImportPageLoaded()
        await this.importFiles.addFiles(['FileUpload/Cat2.txt'])
        await this.b.click({ button: 'Done adding files' })
        await this.b.sleep(1000) // wait for dialog to appear
        await this.b.click({ link: 'Add new field', wait: true })
        await this.b.sendKeys('moo', { tag: 'input', id: 'document-metadata-add-field' })
        await this.b.click('#document-metadata-add-field + button')
        // wait for JS to add field (Jenkins found an error once)
        await this.b.sendKeys('mar', { tag: 'input', name: 'moo', wait: true })
        await this.b.click({ button: 'Import documents' })
        await this.importFiles.waitUntilRedirectToDocumentSet('metadata-test')

        // Check the first document still has 'foo':'bar'
        await this.b.click({ tag: 'h3', contains: 'Cat1.docx' })
        await this.b.sleep(1000) // wait for document to animate in
        await this.b.click({ link: 'Fields', wait: true })
        // wait for metadata to appear
        const value1 = await this.b.getAttribute({ tag: 'input', name: 'foo', wait: true }, 'value')
        expect(value1).to.eq('bar')

        const value2 = await this.b.getAttribute({ tag: 'input', name: 'moo', wait: true }, 'value')
        expect(value2).to.eq('')

        // Check the second document has 'moo':'mar'
        // Avoid a race: how do we detect the second document has loaded? By
        // navigating away from the first document, waiting for it to disappear,
        // navigating to the second, and waiting for it to load.
        await this.b.click({ link: 'Back to list' })
        await this.b.sleep(1000) // FIXME debug, figure out why we need this, then remove it
        await this.b.click({ tag: 'h3', contains: 'Cat2.txt', wait: 'fast' })
        await this.b.sleep(1000) // FIXME debug, figure out why we need this, then remove it
        // wait for metadata to appear
        const value3 = await this.b.getAttribute({ tag: 'input', name: 'foo', wait: true }, 'value')
        expect(value3).to.eq('')

        const value4 = await this.b.getAttribute({ tag: 'input', name: 'moo', wait: true }, 'value')
        expect(value4).to.eq('mar')
      })

      after(async function() {
        await this.documentSets.destroy('metadata-test')
      })
    })

    describe('after uploading files', function() {
      before(async function() {
        await this.importFiles.open()
        await this.importFiles.addFiles([
          'FileUpload/Cat0.pdf',
          'FileUpload/Cat1.docx',
          'FileUpload/Cat2.txt',
          'FileUpload/Cat3.rtf',
          'FileUpload/Cat4.html',
          'FileUpload/Jules1.doc',
          'FileUpload/Jules2.pptx',
          'FileUpload/Jules3.xlsx',
        ])
        await this.importFiles.finish({ name: 'FileUpload' })
      })

      after(async function() {
        await this.documentSets.destroy('FileUpload')
      })

      shouldBehaveLikeATree({
        documents: [
          { type: 'pdf', title: 'Cat0.pdf' },
          { type: 'pdf', title: 'Cat1.docx' },
          { type: 'pdf', title: 'Cat2.txt' },
          { type: 'pdf', title: 'Cat3.rtf' },
          { type: 'pdf', title: 'Cat4.html' },
          { type: 'pdf', title: 'Jules1.doc' },
          { type: 'pdf', title: 'Jules2.pptx' },
          { type: 'pdf', title: 'Jules3.xlsx' },
        ],
        searches: [
          { query: 'chase', nResults: 5 }
        ],
      })
    })

    describe('after splitting a file into pages', function() {
      before(async function() {
        await this.importFiles.open()
        await this.importFiles.addFiles([ 'FileUpload/Cat1.docx' ])
        await this.importFiles.finish({ name: 'Split FileUpload', splitByPage: true })
      })

      after(async function() {
        await this.documentSets.destroy('Split FileUpload')
      })

      shouldBehaveLikeATree({
        documents: [
          { type: 'pdf', title: 'Cat1.docx – page 1' },
          { type: 'pdf', title: 'Cat1.docx – page 2' },
          { type: 'pdf', title: 'Cat1.docx – page 3' },
        ],
        searches: [
          { query: 'face', nResults: 3 }
        ],
      })
    })
  })
})

File: integration-test/spec/file_upload_spec.rb
#!/usr/bin/env ruby

require './spec/spec_helper'

describe 'FileUpload' do
  before do
    @user = admin_session.create_test_user
    page.log_in_as(@user)
  end

  after do
    admin_session.destroy_test_user(@user)
  end

  describe 'after uploading a file with metadata' do
    before do
      page.creating_document_set do
        page.click_on('Upload files', wait: WAIT_LOAD)
        page.assert_selector('.upload-folder-prompt', wait: WAIT_LOAD)
        page.execute_script('document.querySelector(".upload-prompt .invisible-file-input").style.opacity = 1')
        page.attach_file('file', '/app/files/file-upload-spec/Cat0.pdf')
        page.click_on('Done adding files')
        # Wait for focus: that's when the dialog is open
        page.wait_for_javascript_to_return_true('document.querySelector("#import-options-name") === document.activeElement', wait: WAIT_FAST)
        page.fill_in('Document set name', with: 'spec')

        page.click_link('Add new field…')
        page.fill_in('New field name', with: 'foo', wait: WAIT_FAST) # wait for field to appear
        page.click_button('Add')
        page.fill_in('foo', with: 'bar', wait: WAIT_FAST) # wait for field to appear

        page.click_button('Import documents')
      end
    end

    it 'should add metadata to the imported file' do
      page.open_document_in_list_with_name('Cat0.pdf')
      page.waiting_for_css_transitions_on_selector('.document-metadata') do
        page.click_link('Fields')
      end
      assert page.has_field?('foo', with: 'bar', wait: WAIT_FAST)
    end

    it 'should add more metadata in a second import to the same document set' do
      # (this also tests that we can add files to an existing document set)

      # 1. Click "Add Documents"
      page.within('nav') do
        page.click_link('spec')
        page.click_link('Add Documents')
      end

      # 2. Add documents and get to the "Import Options" dialog
      page.assert_selector('.upload-folder-prompt', wait: WAIT_LOAD) # wait for page to load
      page.execute_script('document.querySelector(".upload-prompt .invisible-file-input").style.opacity = 1')
      page.attach_file('file', '/app/files/file-upload-spec/Cat1.docx')
      page.click_on('Done adding files', wait: WAIT_FAST) # wait for field to be enabled
      page.assert_selector('.import-options label', text: 'foo', wait: WAIT_FAST) # Wait for dialog (and field) to appear

      # 3. Set new metadata
      page.click_link('Add new field…')
      page.fill_in('New field name', with: 'moo', wait: WAIT_FAST) # wait for field to appear
      page.click_button('Add')
      page.fill_in('moo', with: 'mar', wait: WAIT_FAST) # wait for field to appear
      page.fill_in('foo', with: 'baz') # fill in the _first_ metadata field with a new value

      # 4. Submit
      page.click_button('Import documents')
      page.finish_creating_document_set(hide_tour: false) # the tour is already hidden

      # 5. Check metadata of original doc (it should not be changed)
      page.open_document_in_list_with_name('Cat0.pdf')
      page.waiting_for_css_transitions_on_selector('.document-metadata') do
        page.click_link('Fields')
      end
      assert page.has_field?('foo', with: 'bar', wait: WAIT_FAST)
      assert page.has_field?('moo', with: '')

      # 6. Check metadata of new doc (it should be what was just entered)
      page.within('#document-current') do
        page.click_link('Next')
      end
      assert page.has_field?('foo', with: 'baz', wait: WAIT_FAST)
      assert page.has_field?('moo', with: 'mar')
    end
  end

  describe 'when splitting a file into pages' do
    before do
      page.creating_document_set do
        page.click_on('Upload files', wait: WAIT_LOAD)
        page.assert_selector('.upload-folder-prompt', wait: WAIT_LOAD)
        page.execute_script('document.querySelector(".upload-prompt .invisible-file-input").style.opacity = 1')
        page.attach_file('file', '/app/files/file-upload-spec/Cat0.pdf')
        page.click_on('Done adding files')
        # Wait for focus: that's when the dialog is open
        page.wait_for_javascript_to_return_true('document.querySelector("#import-options-name") === document.activeElement', wait: WAIT_FAST)
        page.fill_in('Document set name', with: 'spec')
        page.choose('Each page is one document')
        page.choose('Assume documents are already text (faster import)')
        page.click_button('Import documents')
      end
    end

    it 'should create a file per page' do
      page.assert_selector('#document-list h3', text: 'Cat0.pdf – page 1')
      page.assert_selector('#document-list h3', text: 'Cat0.pdf – page 2')
      page.assert_selector('#document-list h3', text: 'Cat0.pdf – page 3')
      page.assert_selector('#document-list h3', text: 'Cat0.pdf – page 4')
    end

    it 'should search all pages' do
      page.search_for_q('face')
      page.assert_selector('#document-list-title h3', text: '4 documents')
    end
  end

  it 'should support all file types' do
    page.create_document_set_from_pdfs_in_folder('files/file-upload-spec')
    for title in %w(Cat0.pdf Cat1.docx Cat2.txt Cat3.rtf Cat4.html Cats.zip/Cat0.pdf Cats.zip/Cat1.docx Image.jpg Image.png Jules1.doc Jules2.pptx Jules3.xlsx)
      page.assert_selector('#document-list h3', text: title)
    end
  end
end
